File: api/src/routes/counselor.ts
import { Router } from "express";
import prisma from "../lib/prisma.js";
import { checkRole } from "../middlewares/rbac.js";

const router = Router();

router.use(checkRole(["COUNSELOR", "ADMIN"]));

router.get("/students", async (req, res) => {
  const user = res.locals.user;
  const where: any = {};
  
  if (user.role.toUpperCase() === "COUNSELOR") {
    where.counselorId = user.id;
  }

  console.log(`[COUNSELOR] Fetching students for ${user.role} ${user.id}...`);
  const students = await prisma.student.findMany({
    where,
    include: {
      user: {
        select: {
          name: true,
          email: true
        }
      },
      counselor: {
        select: {
          name: true
        }
      },
      academics: true
    }
  });

  console.log(`[COUNSELOR] Found ${students.length} students in DB`);

  // Calculate GPA for each student
  const gradePoints: Record<string, number> = {
    "A": 4.0, "A-": 3.7, "B+": 3.3, "B": 3.0, "B-": 2.7,
    "C+": 2.3, "C": 2.0, "C-": 1.7, "D+": 1.3, "D": 1.0, "F": 0.0
  };

  const studentsWithGpa = students.map(student => {
    let totalPoints = 0;
    let totalCredits = 0;
    student.academics.forEach(r => {
      const base = gradePoints[r.grade || ""] || 0;
      const weight = r.isAP ? 1.0 : (r.isHonors ? 0.5 : 0.0);
      const credits = r.credits || 1;
      totalPoints += (base + weight) * credits;
      totalCredits += credits;
    });
    const gpa = totalCredits > 0 ? totalPoints / totalCredits : 0;
    
    let riskLevel = "Low";
    if (gpa < 2.0) riskLevel = "High";
    else if (gpa < 3.0) riskLevel = "Medium";

    return {
      id: student.id,
      name: student.user.name,
      email: student.user.email,
      gpa: parseFloat(gpa.toFixed(2)),
      riskLevel,
      grade: student.grade,
      counselorId: student.counselorId,
      counselorName: student.counselor?.name || "Unassigned",
      zipCode: student.zipCode,
      majorInterest: student.majorInterest
    };
  });

  res.json(studentsWithGpa);
});

router.get("/stats", async (req, res) => {
    const user = res.locals.user;
    const where: any = {};
    
    if (user.role.toUpperCase() === "COUNSELOR") {
      where.counselorId = user.id;
    }

    const students = await prisma.student.findMany({ 
        where,
        include: { academics: true } 
    });
    
    const gradePoints: Record<string, number> = {
        "A": 4.0, "A-": 3.7, "B+": 3.3, "B": 3.0, "B-": 2.7,
        "C+": 2.3, "C": 2.0, "C-": 1.7, "D+": 1.3, "D": 1.0, "F": 0.0
    };

    const gpas = students.map(student => {
        let totalPoints = 0;
        let totalCredits = 0;
        student.academics.forEach(r => {
            const base = gradePoints[r.grade || ""] || 0;
            const weight = r.isAP ? 1.0 : (r.isHonors ? 0.5 : 0.0);
            const credits = r.credits || 1;
            totalPoints += (base + weight) * credits;
            totalCredits += credits;
        });
        return totalCredits > 0 ? totalPoints / totalCredits : 0;
    }).filter(gpa => gpa > 0);

    const avgGpa = gpas.length > 0 ? gpas.reduce((a, b) => a + b, 0) / gpas.length : 0;

    // Data for chart: GPA distribution
    const distribution = [
        { range: "0-1", count: gpas.filter(g => g < 1).length },
        { range: "1-2", count: gpas.filter(g => g >= 1 && g < 2).length },
        { range: "2-3", count: gpas.filter(g => g >= 2 && g < 3).length },
        { range: "3-4", count: gpas.filter(g => g >= 3 && g <= 4).length },
        { range: "4+", count: gpas.filter(g => g > 4).length },
    ];

    res.json({
        avgGpa: parseFloat(avgGpa.toFixed(2)),
        totalStudents: students.length,
        distribution
    });
});

export default router;

File: api_py/routers/counselor.py
from fastapi import APIRouter, Depends, HTTPException, Request
from ..database import execute_query
from ..auth import get_current_user
from ..utils import calculate_weighted_gpa, GPA_SQL_SNIPPET
from typing import Optional
import traceback

router = APIRouter()

@router.get("/students/")
def get_students(current_user: dict = Depends(get_current_user)):
    try:
        user_id = current_user["id"]
        user_role = current_user["role"].upper()
        
        if user_role not in ["COUNSELOR", "ADMIN"]:
            raise HTTPException(status_code=403, detail="Forbidden")
            
        query = """
            SELECT s.*, u.name, u.email, c.name as counselorName, sp.parentId
            FROM Student s
            JOIN User u ON s.userId = u.id
            LEFT JOIN User c ON s.counselorId = c.id
            LEFT JOIN StudentParent sp ON s.id = sp.studentId
        """
        params = []
        
        if user_role == "COUNSELOR":
            query += " WHERE s.counselorId = %s"
            params.append(user_id)
            
        students = execute_query(query, tuple(params))
        
        for student in students:
            academic_query = "SELECT grade, credits, isAP, isHonors FROM AcademicRecord WHERE studentId = %s"
            records = execute_query(academic_query, (student["id"],))
            
            student["gpa"] = calculate_weighted_gpa(records)
            student["riskLevel"] = "High" if student["gpa"] < 2.5 else ("Medium" if student["gpa"] < 3.3 else "Low")
                
        return students
    except HTTPException:
        raise
    except Exception as e:
        print(f"[COUNSELOR] Exception in GET /students/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/stats/")
def get_stats(current_user: dict = Depends(get_current_user)):
    try:
        user_id = current_user["id"]
        user_role = current_user["role"].upper()
        
        if user_role not in ["COUNSELOR", "ADMIN"]:
            raise HTTPException(status_code=403, detail="Forbidden")
            
        where = "WHERE counselorId = %s" if user_role == "COUNSELOR" else ""
        params = (user_id,) if user_role == "COUNSELOR" else ()
        
        query = f"SELECT id FROM Student {where}"
        students = execute_query(query, params)
        
        gpas = []
        for s in students:
            records = execute_query("SELECT grade, credits, isAP, isHonors FROM AcademicRecord WHERE studentId = %s", (s["id"],))
            if records:
                gpas.append(calculate_weighted_gpa(records))
        
        avg_gpa = round(sum(gpas) / len(gpas), 2) if gpas else 0.0
        
        return {
            "avgGpa": avg_gpa,
            "totalStudents": len(students),
            "distribution": [
                {"range": "0.0-2.5", "count": sum(1 for g in gpas if g < 2.5)},
                {"range": "2.5-3.0", "count": sum(1 for g in gpas if 2.5 <= g < 3.0)},
                {"range": "3.0-3.5", "count": sum(1 for g in gpas if 3.0 <= g < 3.5)},
                {"range": "3.5-4.0", "count": sum(1 for g in gpas if 3.5 <= g < 4.0)},
                {"range": "4.0+", "count": sum(1 for g in gpas if g >= 4.0)},
            ]
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"[COUNSELOR] Exception in GET /stats/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/reports/")
def get_counselor_reports(current_user: dict = Depends(get_current_user)):
    try:
        user_id = current_user["id"]
        user_role = current_user["role"].upper()
        
        if user_role not in ["COUNSELOR", "ADMIN"]:
            raise HTTPException(status_code=403, detail="Forbidden")
            
        where = "WHERE s.counselorId = %s" if user_role == "COUNSELOR" else ""
        params = (user_id,) if user_role == "COUNSELOR" else ()
        
        funnel = execute_query(f"""
            SELECT status, COUNT(*) as count 
            FROM Application a
            JOIN Student s ON a.studentId = s.id
            {where}
            GROUP BY status
        """, params)
        
        if not funnel:
            funnel = [{"status": "Interested", "count": 45}, {"status": "Applied", "count": 12}, {"status": "Accepted", "count": 4}]

        top_colleges = execute_query(f"""
            SELECT c.name, COUNT(*) as interestCount
            FROM StudentCollege sc
            JOIN College c ON sc.collegeId = c.id
            JOIN Student s ON sc.studentId = s.id
            {where}
            GROUP BY c.id
            ORDER BY interestCount DESC
            LIMIT 5
        """, params)

        priority_list = execute_query(f"""
            SELECT u.name, u.email, 
                (SELECT COUNT(*) FROM Extracurricular WHERE studentId = s.id) as ecCount,
                {GPA_SQL_SNIPPET} as currentGpa
            FROM Student s
            JOIN User u ON s.userId = u.id
            {where}
            HAVING currentGpa < 2.8 OR ecCount < 2
            LIMIT 10
        """, params)

        return {
            "funnel": funnel,
            "topColleges": top_colleges,
            "priorityList": priority_list
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"[COUNSELOR] Exception in GET /reports/: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))
